tsl/src/nodes/decompress_chunk/sorted_merge.c

```diff
@@ -118,7 +118,18 @@ decompress_batch_open_next_batch(DecompressChunkState *chunk_state)
 
 			chunk_state->most_recent_batch = batch_state_id;
 
-			return;
+			/* In decompress_sorted_merge_get_next_tuple it is determined how many batches
+			 * need to be opened currently to perform a sorted merge. This is done by
+			 * checking if the first tuple from the last opened batch is larger than the
+			 * last returned tuple.
+			 *
+			 * If a filter removes the first tuple, the first into the heap inserted
+			 * tuple from this batch can no longer be used to perform the check.
+			 * Therefore, we must continue opening additional batches until the condition
+			 * is met.
+			 */
+			if (first_tuple_returned)
```

Comment:
I think you might want to open all batches; up to the current `peek` element - or not?
...or you could probably move the filter out from applying it during read right away - and retain the old logic
...or also add the chunk boundaries to the heap - and always use the `max` level to pull in new batches; possibly pull in by min for the 1st element - but the `peek` based approach is probably simpler

Reply:
This is correct, if the first element is removed from the batch, the existing look-ahead logic can not use the returned tuple to decide if we have to open more batches.

This PR would exclude batches from the look-ahead logic if the first tuple was filtered. This could lead to a situation where we open more batches than needed. 

Returning the first tuple from the batch, regardless if it satisfies the filter condition or not, could indeed help to reduce the number of open batches. However, this would also require some extra logic, which I wanted to avoid because this algorithm already has many special cases and I didn't want to make it more complicated.

Reply:
if a filter has `p` probability to filter out a row; the chance to have a batch with at least one element matching; but not the first is:
* `(first row is filtered) * (at least 1 match in remaining)`
* `(first row is filtered) * (1-(all rows filtered))`
* which is: `p * (1-p^999)`
* this function can get very close to `1`  https://www.wolframalpha.com/input?i=max%28p+*+%281-p%5E999%29%29
* which means it will just open up all batches; but because of the selectivity it will most likely also have 100% excluded batches

but the best would be to do something with it later...dual-heap or something else; doesn't matter.